transport/peer.go

```diff
@@ -0,0 +1,60 @@
+package transport
+
+//go:generate mockgen -destination=transporttest/peer.go -package=transporttest go.uber.org/yarpc/transport PeerIdentifier,Peer,SubscribablePeer,PeerSubscriber
+
+// PeerStatus maintains information about the Peer's connection state
+type PeerStatus string
+
+const (
+	// PeerAvailable indicates the Peer is available for requests
+	PeerAvailable PeerStatus = "available"
+	// PeerConnecting indicates the Peer is in the process of connecting
+	PeerConnecting PeerStatus = "connecting"
+	// PeerUnavailable indicates the Peer is unavailable for requests
+	PeerUnavailable PeerStatus = "unavailable"
+)
+
+/*
+PeerIdentifier is able to uniquely identify a peer (e.g. hostport)
+*/
+type PeerIdentifier interface {
+	Identifier() string
+}
+
+/*
+Peer is a level on top of PeerIdentifier.  It should be created by a PeerAgent so we
+can maintain multiple references to the same downstream peer (e.g. hostport).  This is
+useful for load balancing requests to downstream services.
+*/
+type Peer interface {
+	PeerIdentifier
+
+	GetStatus() PeerStatus // Get the status of the Peer
+	GetAgent() PeerAgent   // Get the PeerAgent which owns this Peer
+	Pending() int          // The number of pending requests on this peer
```

Comment:
sounds good, I like the idea of the status struct because we should be able to add more parameters to it without breaking interfaces